File: Recursion/Lecture 13/CodeStudio/Combination Sum II.cpp

#include <bits/stdc++.h>
using namespace std;

void help(int i, vector<int> &arr, int n, vector<int> &subSet, vector<vector<int>> &powerSet, int sum, int target)
{
    // base condition
    // when sum == target this means that we have reached a Valid Combination thus include it into the vector<vector<int>> powerSet
    if (sum == target)
    {
        powerSet.push_back(subSet);
        return;
    }

    // base condition
    // when sum > target this means that we are exceeding the limit of the required sum (target) therefore simply return
    if (sum > target)
        return;

    // base condition
    // if current Index (i) is exceeding the array size then return
    if (i == n)
        return;

    // pick the ith element
    subSet.push_back(arr[i]);
    sum += arr[i];

    // ask recursion to do the remaining task
    help(i + 1, arr, n, subSet, powerSet, sum, target);

    // backtrack & undo the changes that were made while picking the ith element
    subSet.pop_back();
    sum -= arr[i];

    // loop to ensure that we don't pick any other occurrences of ith element
    while (i + 1 < arr.size() && arr[i] == arr[i + 1])
        i++;

    // skip the ith element
    help(i + 1, arr, n, subSet, powerSet, sum, target);
}

vector<vector<int>> combinationSum2(vector<int> &arr, int n, int target)
{
    vector<int> subSet;
    vector<vector<int>> powerSet;
    int sum = 0;

    // as we have to bring all the similar elements together therefore we need to sort arr[]
    // now as all the similar elements comes together therefore we will be able to ignore all the duplicate occurences of elements
    sort(arr.begin(), arr.end());
    help(0, arr, n, subSet, powerSet, sum, target);
    return powerSet;
}

/*
    time complexity : O(2^n)
    space complexity : O(n)
*/

File: Recursion/Lecture 13/LeetCode/Combination Sum II.java
/* This question is exactly similar to Combination Sum -I. 
The only difference is that we have duplicates inside the given array arr[]
We need to skip the duplicate elements. Else they would generate Duplicate Combinations which we don't want
To skip the Duplicate Elements, we will sort arr[] and use a While Loop to skip all the Duplicate Elements
*/

class Solution 
{
    public void help(int i, int[] arr, int n, List<Integer> subSet, List<List<Integer>> powerSet, int sum, int target) 
    {
    
        // If sum is equal to target, we have reached a Valid Combination

        if(sum == target) 
        {
            powerSet.add(new ArrayList(subSet)) ;
            return;
        }

        // If at any moment, sum becomes greater than target, we don't need to proceed further

        if(sum > target) return ;

        // If we reach the end of arr[], we cannot go any further so we return back

        if(i == n) return;

        // Include the i-th Element into our Subset & Sum

        subSet.add(arr[i]) ;
        sum += arr[i] ;

        // Ask recursion to do rest of the task
        help(i + 1, arr, n, subSet, powerSet, sum, target) ;

        // Backtrack and undo the change we have done

        subSet.remove(subSet.size()-1);
        sum -= arr[i];

        // Use the While Loop to skip all the duplicate occurrences of i-th Element

        while(i + 1 < arr.length && arr[i] == arr[i + 1]) i++ ;

        // Don't pick the i-th Element and ask recursion to do rest of the task

        help(i + 1, arr, n, subSet, powerSet, sum, target) ;
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) 
    {
        List<Integer> subSet = new ArrayList<>() ;
        List<List<Integer>> powerSet = new ArrayList<>() ;
        int sum = 0;
        int n = candidates.length ;
        Arrays.sort(candidates) ;

        help(0, candidates, n, subSet, powerSet, sum, target) ;

        return powerSet ;
    }
}

/* 
Time Complexity: O(2^N)
Space Complexity: O(N)
*/
